# ڈیٹا کی تیاری

[اصل نوٹ بک کا ماخذ *ڈیٹا سائنس: ڈیٹا سائنس کے لیے مشین لرننگ کا تعارف، پائتھون اور مشین لرننگ اسٹوڈیو از لی اسٹاٹ*](https://github.com/leestott/intro-Datascience/blob/master/Course%20Materials/4-Cleaning_and_Manipulating-Reference.ipynb)

## `DataFrame` کی معلومات کا جائزہ لینا

> **سیکھنے کا مقصد:** اس ذیلی حصے کے اختتام تک، آپ pandas DataFrames میں محفوظ کردہ ڈیٹا کے بارے میں عمومی معلومات حاصل کرنے میں مہارت حاصل کر لیں گے۔

جب آپ نے اپنا ڈیٹا pandas میں لوڈ کر لیا ہو، تو زیادہ امکان ہے کہ یہ `DataFrame` میں ہوگا۔ لیکن اگر آپ کے `DataFrame` میں موجود ڈیٹا سیٹ میں 60,000 قطاریں اور 400 کالم ہوں، تو آپ کام شروع کرنے کے لیے اس کا اندازہ کیسے لگائیں گے؟ خوش قسمتی سے، pandas کچھ آسان ٹولز فراہم کرتا ہے جو `DataFrame` کے بارے میں مجموعی معلومات کے ساتھ ساتھ ابتدائی اور آخری چند قطاروں کو جلدی دیکھنے میں مدد کرتے ہیں۔

اس فعالیت کو دریافت کرنے کے لیے، ہم Python کی scikit-learn لائبریری درآمد کریں گے اور ایک مشہور ڈیٹا سیٹ استعمال کریں گے جسے ہر ڈیٹا سائنسدان نے سینکڑوں بار دیکھا ہے: برطانوی حیاتیات دان رونالڈ فشر کا *Iris* ڈیٹا سیٹ، جو انہوں نے 1936 کے اپنے مقالے "The use of multiple measurements in taxonomic problems" میں استعمال کیا تھا:


In [1]:
import pandas as pd
from sklearn.datasets import load_iris

iris = load_iris()
iris_df = pd.DataFrame(data=iris['data'], columns=iris['feature_names'])

### `DataFrame.shape`
ہم نے Iris Dataset کو متغیر `iris_df` میں لوڈ کیا ہے۔ ڈیٹا میں گہرائی سے جانے سے پہلے یہ جاننا مفید ہوگا کہ ہمارے پاس کتنے ڈیٹا پوائنٹس ہیں اور ڈیٹا سیٹ کا مجموعی سائز کیا ہے۔ یہ دیکھنا فائدہ مند ہے کہ ہم کس حجم کے ڈیٹا سے نمٹ رہے ہیں۔


In [2]:
iris_df.shape

(150, 4)

تو، ہم 150 قطاروں اور 4 کالمز کے ڈیٹا کے ساتھ کام کر رہے ہیں۔ ہر قطار ایک ڈیٹا پوائنٹ کی نمائندگی کرتی ہے اور ہر کالم ڈیٹا فریم سے متعلق ایک خصوصیت کی نمائندگی کرتا ہے۔ تو بنیادی طور پر، یہاں 150 ڈیٹا پوائنٹس ہیں جن میں ہر ایک کے ساتھ 4 خصوصیات موجود ہیں۔

`shape` یہاں ڈیٹا فریم کی ایک خصوصیت ہے نہ کہ ایک فنکشن، اسی لیے یہ ایک جوڑے قوسین پر ختم نہیں ہوتی۔


### `DataFrame.columns`
اب ہم ڈیٹا کے 4 کالمز کی طرف بڑھتے ہیں۔ ہر کالم بالکل کس چیز کی نمائندگی کرتا ہے؟ `columns` ایٹریبیوٹ ہمیں ڈیٹا فریم میں کالمز کے نام فراہم کرے گا۔


In [3]:
iris_df.columns

Index(['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)',
       'petal width (cm)'],
      dtype='object')

جیسا کہ ہم دیکھ سکتے ہیں، یہاں چار (4) کالم ہیں۔ `columns` خصوصیت ہمیں کالمز کے نام بتاتی ہے اور بنیادی طور پر کچھ اور نہیں۔ یہ خصوصیت اس وقت اہمیت اختیار کرتی ہے جب ہم یہ شناخت کرنا چاہتے ہیں کہ ایک ڈیٹاسیٹ میں کون سی خصوصیات شامل ہیں۔


### `DataFrame.info`
ڈیٹا کی مقدار (`shape` attribute کے ذریعے دی گئی) اور فیچرز یا کالمز کے نام (`columns` attribute کے ذریعے دیے گئے) ہمیں ڈیٹاسیٹ کے بارے میں کچھ معلومات فراہم کرتے ہیں۔ اب، ہم ڈیٹاسیٹ کو مزید گہرائی میں سمجھنا چاہیں گے۔ `DataFrame.info()` فنکشن اس کے لیے کافی مفید ہے۔


In [4]:
iris_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 4 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
dtypes: float64(4)
memory usage: 4.8 KB


یہاں سے، ہم چند مشاہدات کر سکتے ہیں:
1. ہر کالم کا DataType: اس ڈیٹا سیٹ میں، تمام ڈیٹا 64-bit floating-point نمبروں کی صورت میں محفوظ ہے۔
2. Non-Null ویلیوز کی تعداد: null ویلیوز سے نمٹنا ڈیٹا کی تیاری میں ایک اہم قدم ہے۔ اس پر بعد میں نوٹ بک میں کام کیا جائے گا۔


### DataFrame.describe()
فرض کریں کہ ہمارے ڈیٹا سیٹ میں بہت سا عددی ڈیٹا موجود ہے۔ یک متغیری شماریاتی حسابات جیسے اوسط، درمیانی قدر، چوتھائی وغیرہ ہر کالم پر الگ الگ کیے جا سکتے ہیں۔ `DataFrame.describe()` فنکشن ہمیں ڈیٹا سیٹ کے عددی کالمز کا شماریاتی خلاصہ فراہم کرتا ہے۔


In [5]:
iris_df.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
count         150.000000        150.000000         150.000000        150.000000
mean            5.843333          3.057333           3.758000          1.199333
std             0.828066          0.435866           1.765298          0.762238
min             4.300000          2.000000           1.000000          0.100000
25%             5.100000          2.800000           1.600000          0.300000
50%             5.800000          3.000000           4.350000          1.300000
75%             6.400000          3.300000           5.100000          1.800000
max             7.900000          4.400000           6.900000          2.500000

اوپر دیا گیا نتیجہ ہر کالم کے کل ڈیٹا پوائنٹس، اوسط، معیاری انحراف، کم از کم، نچلا چوتھائی (25٪)، درمیانی (50٪)، اوپری چوتھائی (75٪) اور زیادہ سے زیادہ قدر کو ظاہر کرتا ہے۔


### `DataFrame.head`
تمام مذکورہ فنکشنز اور خصوصیات کے ساتھ، ہمیں ڈیٹا سیٹ کا ایک اعلیٰ سطحی جائزہ مل گیا ہے۔ ہمیں معلوم ہے کہ کتنے ڈیٹا پوائنٹس موجود ہیں، کتنی خصوصیات ہیں، ہر خصوصیت کا ڈیٹا ٹائپ کیا ہے اور ہر خصوصیت کے غیر خالی (non-null) ویلیوز کی تعداد کیا ہے۔

اب وقت ہے کہ خود ڈیٹا کو دیکھیں۔ آئیے دیکھتے ہیں کہ ہمارے `DataFrame` کی ابتدائی چند قطاریں (ابتدائی چند ڈیٹا پوائنٹس) کیسی نظر آتی ہیں:


In [6]:
iris_df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

جیسا کہ یہاں آؤٹ پٹ میں، ہم ڈیٹاسیٹ کے پانچ (5) اندراجات دیکھ سکتے ہیں۔ اگر ہم بائیں طرف کے انڈیکس کو دیکھیں، تو ہمیں پتہ چلتا ہے کہ یہ پہلی پانچ قطاریں ہیں۔


### مشق:

اوپر دیے گئے مثال سے یہ واضح ہے کہ، ڈیفالٹ کے طور پر، `DataFrame.head` ایک `DataFrame` کی پہلی پانچ قطاریں واپس کرتا ہے۔ نیچے دیے گئے کوڈ سیل میں، کیا آپ ایک طریقہ تلاش کر سکتے ہیں جس سے پانچ سے زیادہ قطاریں دکھائی جا سکیں؟


In [7]:
# Hint: Consult the documentation by using iris_df.head?

### `DataFrame.tail`
ڈیٹا کو دیکھنے کا ایک اور طریقہ آخر سے ہو سکتا ہے (شروع کے بجائے)۔ `DataFrame.head` کا الٹ `DataFrame.tail` ہے، جو `DataFrame` کی آخری پانچ قطاریں واپس کرتا ہے:


In [8]:
iris_df.tail()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

عملی طور پر، یہ فائدہ مند ہوتا ہے کہ آپ آسانی سے `DataFrame` کی ابتدائی چند قطاروں یا آخری چند قطاروں کا جائزہ لے سکیں، خاص طور پر جب آپ ترتیب شدہ ڈیٹا سیٹس میں غیر معمولی اقدار تلاش کر رہے ہوں۔

کوڈ کی مثالوں کی مدد سے دکھائی گئی تمام فنکشنز اور خصوصیات ہمیں ڈیٹا کی جھلک اور احساس حاصل کرنے میں مدد دیتی ہیں۔

> **خلاصہ:** صرف `DataFrame` میں موجود معلومات کے میٹا ڈیٹا یا اس کی ابتدائی اور آخری چند قدروں کو دیکھ کر ہی آپ فوری طور پر اس ڈیٹا کے حجم، شکل، اور مواد کے بارے میں ایک اندازہ لگا سکتے ہیں جس کے ساتھ آپ کام کر رہے ہیں۔


### گمشدہ ڈیٹا
آئیے گمشدہ ڈیٹا کے بارے میں بات کرتے ہیں۔ گمشدہ ڈیٹا تب ہوتا ہے جب کچھ کالمز میں کوئی ویلیو محفوظ نہ ہو۔

آئیے ایک مثال لیتے ہیں: فرض کریں کوئی شخص اپنے وزن کے بارے میں حساس ہے اور سروے میں وزن کا خانہ نہیں بھرتا۔ تو اس شخص کے لیے وزن کی ویلیو گمشدہ ہوگی۔

زیادہ تر وقت، حقیقی دنیا کے ڈیٹا سیٹس میں گمشدہ ویلیوز ہوتی ہیں۔

**پینڈاز گمشدہ ڈیٹا کو کیسے ہینڈل کرتا ہے**

پینڈاز گمشدہ ویلیوز کو دو طریقوں سے ہینڈل کرتا ہے۔ پہلا طریقہ آپ پہلے دیکھ چکے ہیں: `NaN`، یا "Not a Number"۔ یہ دراصل ایک خاص ویلیو ہے جو IEEE فلوٹنگ پوائنٹ اسپیسفیکیشن کا حصہ ہے اور یہ صرف گمشدہ فلوٹنگ پوائنٹ ویلیوز کو ظاہر کرنے کے لیے استعمال ہوتی ہے۔

فلوٹس کے علاوہ گمشدہ ویلیوز کے لیے، پینڈاز Python کے `None` آبجیکٹ کا استعمال کرتا ہے۔ اگرچہ یہ تھوڑا الجھن پیدا کر سکتا ہے کہ آپ کو دو مختلف قسم کی ویلیوز ملیں گی جو بنیادی طور پر ایک ہی بات کہتی ہیں، لیکن اس ڈیزائن کے انتخاب کے پیچھے مضبوط پروگرامنگ وجوہات ہیں۔ عملی طور پر، یہ طریقہ پینڈاز کو زیادہ تر کیسز کے لیے ایک اچھا حل فراہم کرنے کے قابل بناتا ہے۔ اس کے باوجود، `None` اور `NaN` دونوں میں کچھ پابندیاں ہیں جنہیں آپ کو ذہن میں رکھنا ہوگا کہ انہیں کیسے استعمال کیا جا سکتا ہے۔


### `None`: غیر فلوٹ گمشدہ ڈیٹا
چونکہ `None` Python سے آتا ہے، اسے NumPy اور pandas arrays میں استعمال نہیں کیا جا سکتا جو ڈیٹا ٹائپ `'object'` کے نہیں ہیں۔ یاد رکھیں، NumPy arrays (اور pandas میں ڈیٹا اسٹرکچرز) صرف ایک قسم کے ڈیٹا پر مشتمل ہو سکتے ہیں۔ یہی چیز انہیں بڑے پیمانے پر ڈیٹا اور حسابی کام کے لیے زبردست طاقت دیتی ہے، لیکن یہ ان کی لچک کو بھی محدود کرتی ہے۔ ایسے arrays کو "کم از کم مشترکہ عنصر" میں تبدیل کرنا پڑتا ہے، یعنی وہ ڈیٹا ٹائپ جو array میں موجود ہر چیز کو شامل کر سکے۔ جب array میں `None` موجود ہو، تو اس کا مطلب ہے کہ آپ Python objects کے ساتھ کام کر رہے ہیں۔

اسے عملی طور پر دیکھنے کے لیے، درج ذیل مثال کے array پر غور کریں (اس کا `dtype` نوٹ کریں):


In [9]:
import numpy as np

example1 = np.array([2, None, 6, 8])
example1

array([2, None, 6, 8], dtype=object)

ڈیٹا ٹائپس کے اپکاسٹ ہونے کی حقیقت کے ساتھ دو اثرات جڑے ہوتے ہیں۔ پہلا یہ کہ آپریشنز انٹرپریٹڈ پائتھون کوڈ کی سطح پر انجام دیے جائیں گے، نہ کہ کمپائلڈ NumPy کوڈ کی سطح پر۔ بنیادی طور پر، اس کا مطلب یہ ہے کہ اگر `Series` یا `DataFrames` میں `None` موجود ہو تو ان پر کیے جانے والے آپریشنز سست ہوں گے۔ اگرچہ آپ کو یہ کارکردگی کا فرق شاید محسوس نہ ہو، لیکن بڑے ڈیٹا سیٹس کے لیے یہ مسئلہ بن سکتا ہے۔

دوسرا اثر پہلے اثر سے جڑا ہوا ہے۔ کیونکہ `None` بنیادی طور پر `Series` یا `DataFrame`s کو واپس عام پائتھون کی دنیا میں لے جاتا ہے، اس لیے اگر آپ NumPy/pandas کے ایگریگیشن فنکشنز جیسے `sum()` یا `min()` کو ایسے arrays پر استعمال کریں جن میں ``None`` موجود ہو، تو عام طور پر ایک ایرر پیدا ہوگا:


In [10]:
example1.sum()

TypeError: ignored

**اہم نکتہ**: عددی اقدار اور `None` کے درمیان جمع (اور دیگر عملیات) غیر معین ہے، جو ایسے ڈیٹا سیٹس کے ساتھ کام کرنے کی صلاحیت کو محدود کر سکتا ہے جن میں یہ موجود ہوں۔


### `NaN`: گمشدہ فلوٹ ویلیوز

`None` کے برعکس، NumPy (اور اسی طرح pandas) اپنے تیز، ویکٹرائزڈ آپریشنز اور ufuncs کے لیے `NaN` کو سپورٹ کرتا ہے۔ بری خبر یہ ہے کہ `NaN` پر کی جانے والی کوئی بھی حسابی کارروائی ہمیشہ `NaN` ہی دیتی ہے۔ مثال کے طور پر:


In [11]:
np.nan + 1

nan

In [12]:
np.nan * 0

nan

اچھی خبر: ایسے arrays پر aggregations چلانے سے جن میں `NaN` ہو، errors ظاہر نہیں ہوتے۔ بری خبر: نتائج یکساں طور پر مفید نہیں ہوتے:


In [13]:
example2 = np.array([2, np.nan, 6, 8]) 
example2.sum(), example2.min(), example2.max()

(nan, nan, nan)

### ورزش:


In [11]:
# What happens if you add np.nan and None together?


یاد رکھیں: `NaN` صرف گمشدہ فلوٹنگ پوائنٹ قدروں کے لیے ہے؛ عددی، سٹرنگز یا بولینز کے لیے کوئی `NaN` متبادل نہیں ہے۔


### `NaN` اور `None`: pandas میں null ویلیوز

اگرچہ `NaN` اور `None` کا رویہ کچھ مختلف ہو سکتا ہے، pandas کو اس طرح بنایا گیا ہے کہ وہ انہیں ایک دوسرے کے ساتھ استعمال کر سکے۔ اس بات کو سمجھنے کے لیے، ایک integers کی `Series` پر غور کریں:


In [15]:
int_series = pd.Series([1, 2, 3], dtype=int)
int_series

0    1
1    2
2    3
dtype: int64

### ورزش:


In [16]:
# Now set an element of int_series equal to None.
# How does that element show up in the Series?
# What is the dtype of the Series?


پینڈاز میں `Series` اور `DataFrame`s کے ڈیٹا کی یکسانیت قائم کرنے کے لیے ڈیٹا ٹائپس کو اپ کاسٹ کرنے کے عمل میں، پینڈاز آسانی سے گمشدہ اقدار کو `None` اور `NaN` کے درمیان تبدیل کر سکتا ہے۔ اس ڈیزائن فیچر کی وجہ سے، پینڈاز میں `None` اور `NaN` کو "null" کی دو مختلف اقسام کے طور پر سوچنا مفید ہو سکتا ہے۔ درحقیقت، پینڈاز میں گمشدہ اقدار سے نمٹنے کے لیے استعمال ہونے والے کچھ بنیادی طریقے اس خیال کو اپنے ناموں میں ظاہر کرتے ہیں:

- `isnull()`: گمشدہ اقدار کی نشاندہی کرنے کے لیے ایک Boolean ماسک تیار کرتا ہے
- `notnull()`: `isnull()` کے برعکس کام کرتا ہے
- `dropna()`: ڈیٹا کا فلٹر شدہ ورژن واپس کرتا ہے
- `fillna()`: ڈیٹا کی ایک کاپی واپس کرتا ہے جس میں گمشدہ اقدار کو بھر دیا گیا ہو یا ان کی جگہ متبادل دی گئی ہو

یہ طریقے سیکھنا اور ان کے ساتھ مہارت حاصل کرنا بہت ضروری ہے، تو آئیے ان میں سے ہر ایک کو تفصیل سے سمجھتے ہیں۔


### خالی اقدار کا پتہ لگانا

اب جب کہ ہم نے گمشدہ اقدار کی اہمیت کو سمجھ لیا ہے، ہمیں اپنے ڈیٹا سیٹ میں ان کا پتہ لگانا ہوگا، اس سے نمٹنے سے پہلے۔
`isnull()` اور `notnull()` دونوں خالی ڈیٹا کا پتہ لگانے کے لیے آپ کے بنیادی طریقے ہیں۔ دونوں آپ کے ڈیٹا پر Boolean ماسک واپس کرتے ہیں۔


In [17]:
example3 = pd.Series([0, np.nan, '', None])

In [18]:
example3.isnull()

0    False
1     True
2    False
3     True
dtype: bool

کیا آپ کو نتیجہ دیکھ کر کوئی حیرت ہوئی؟ اگرچہ `0` ایک حسابی خالی قدر ہے، یہ پھر بھی ایک بالکل درست عدد ہے اور pandas اسے اسی طرح سمجھتا ہے۔ `''` تھوڑا زیادہ پیچیدہ ہے۔ اگرچہ ہم نے اسے سیکشن 1 میں خالی سٹرنگ ویلیو کی نمائندگی کے لیے استعمال کیا تھا، یہ پھر بھی ایک سٹرنگ آبجیکٹ ہے اور pandas کے لحاظ سے خالی قدر کی نمائندگی نہیں کرتا۔

اب، آئیے اس کو الٹ کر دیکھتے ہیں اور ان طریقوں کو اس طرح استعمال کرتے ہیں جیسے آپ انہیں عملی طور پر استعمال کریں گے۔ آپ Boolean ماسک کو براہ راست ``Series`` یا ``DataFrame`` انڈیکس کے طور پر استعمال کر سکتے ہیں، جو الگ تھلگ خالی (یا موجود) قدروں کے ساتھ کام کرنے کی کوشش کرتے وقت مفید ہو سکتا ہے۔

اگر ہم خالی قدروں کی کل تعداد جاننا چاہتے ہیں، تو ہم صرف `isnull()` طریقے سے پیدا ہونے والے ماسک پر ایک سم کر سکتے ہیں۔


In [19]:
example3.isnull().sum()

2

### ورزش:


In [20]:
# Try running example3[example3.notnull()].
# Before you do so, what do you expect to see?


**اہم نکتہ**: دونوں `isnull()` اور `notnull()` طریقے ڈیٹا فریمز میں استعمال کرنے پر مشابہ نتائج پیدا کرتے ہیں: یہ نتائج اور ان نتائج کے انڈیکس کو دکھاتے ہیں، جو آپ کے ڈیٹا کے ساتھ کام کرتے وقت آپ کی بہت مدد کریں گے۔


### گمشدہ ڈیٹا سے نمٹنا

> **سیکھنے کا مقصد:** اس حصے کے اختتام تک، آپ کو معلوم ہونا چاہیے کہ ڈیٹا فریمز سے null ویلیوز کو کب اور کیسے تبدیل یا ہٹایا جائے۔

مشین لرننگ ماڈلز خود گمشدہ ڈیٹا کو نہیں سنبھال سکتے۔ لہذا، ماڈل میں ڈیٹا دینے سے پہلے، ہمیں ان گمشدہ ویلیوز کو حل کرنا ہوگا۔

گمشدہ ڈیٹا کو سنبھالنے کا طریقہ اپنے ساتھ باریک تجارتی پہلو لے کر آتا ہے، جو آپ کے حتمی تجزیے اور حقیقی دنیا کے نتائج کو متاثر کر سکتا ہے۔

گمشدہ ڈیٹا سے نمٹنے کے بنیادی طور پر دو طریقے ہیں:

1.   گمشدہ ویلیو والی قطار کو حذف کریں
2.   گمشدہ ویلیو کو کسی اور ویلیو سے تبدیل کریں

ہم ان دونوں طریقوں اور ان کے فوائد و نقصانات کو تفصیل سے زیر بحث لائیں گے۔


### خالی قدروں کو ہٹانا

جو ڈیٹا ہم اپنے ماڈل کو فراہم کرتے ہیں، اس کی مقدار براہ راست اس کی کارکردگی پر اثر ڈالتی ہے۔ خالی قدروں کو ہٹانے کا مطلب یہ ہے کہ ہم ڈیٹا پوائنٹس کی تعداد کم کر رہے ہیں، اور اس طرح ڈیٹاسیٹ کا سائز بھی کم ہو رہا ہے۔ لہٰذا، یہ مشورہ دیا جاتا ہے کہ جب ڈیٹاسیٹ کافی بڑا ہو تو خالی قدروں والی قطاروں کو ہٹا دیا جائے۔

ایک اور صورت یہ ہو سکتی ہے کہ کسی خاص قطار یا کالم میں بہت زیادہ خالی قدریں ہوں۔ تب، انہیں ہٹایا جا سکتا ہے کیونکہ وہ ہماری تجزیے میں زیادہ قدر نہیں ڈالیں گے کیونکہ اس قطار/کالم کے زیادہ تر ڈیٹا غائب ہیں۔

خالی قدروں کی شناخت سے آگے، pandas ایک آسان طریقہ فراہم کرتا ہے `Series` اور `DataFrame`s سے خالی قدروں کو ہٹانے کا۔ اس کو عملی طور پر دیکھنے کے لیے، آئیے `example3` پر واپس چلتے ہیں۔ `DataFrame.dropna()` فنکشن خالی قدروں والی قطاروں کو ہٹانے میں مدد کرتا ہے۔


In [21]:
example3 = example3.dropna()
example3

0    0
2     
dtype: object

نوٹ کریں کہ یہ آپ کے `example3[example3.notnull()]` کے آؤٹ پٹ کی طرح نظر آنا چاہیے۔ یہاں فرق یہ ہے کہ، صرف ماسک شدہ اقدار پر انڈیکسنگ کرنے کے بجائے، `dropna` نے `Series` `example3` سے وہ گمشدہ اقدار ہٹا دی ہیں۔

چونکہ DataFrames دو جہتوں کے حامل ہوتے ہیں، یہ ڈیٹا کو ہٹانے کے لیے مزید اختیارات فراہم کرتے ہیں۔


In [22]:
example4 = pd.DataFrame([[1,      np.nan, 7], 
                         [2,      5,      8], 
                         [np.nan, 6,      9]])
example4

0    1  2
0  1.0  NaN  7
1  2.0  5.0  8
2  NaN  6.0  9

(کیا آپ نے دیکھا کہ pandas نے `NaN`s کو ایڈجسٹ کرنے کے لیے دو کالمز کو floats میں تبدیل کر دیا؟)

آپ `DataFrame` سے ایک واحد ویلیو کو نہیں ہٹا سکتے، اس لیے آپ کو مکمل قطاریں یا کالمز ہٹانے ہوں گے۔ جو کام آپ کر رہے ہیں اس پر منحصر ہے، آپ ایک یا دوسرا انتخاب کر سکتے ہیں، اور اسی لیے pandas آپ کو دونوں کے لیے اختیارات دیتا ہے۔ کیونکہ ڈیٹا سائنس میں، کالمز عام طور پر متغیرات کی نمائندگی کرتے ہیں اور قطاریں مشاہدات کی، آپ زیادہ تر ڈیٹا کی قطاریں ہٹانے کو ترجیح دیں گے؛ `dropna()` کی ڈیفالٹ سیٹنگ یہ ہے کہ وہ تمام قطاریں ہٹا دے جو کسی بھی null ویلیو پر مشتمل ہوں:


In [23]:
example4.dropna()

0    1  2
1  2.0  5.0  8

اگر ضروری ہو، تو آپ کالمز سے NA ویلیوز کو ہٹا سکتے ہیں۔ ایسا کرنے کے لیے `axis=1` استعمال کریں:


In [24]:
example4.dropna(axis='columns')

2
0  7
1  8
2  9

نوٹ کریں کہ یہ بہت سا ڈیٹا ختم کر سکتا ہے جو آپ رکھنا چاہتے ہیں، خاص طور پر چھوٹے ڈیٹا سیٹس میں۔ اگر آپ صرف وہ قطاریں یا کالم ختم کرنا چاہتے ہیں جن میں کئی یا تمام null ویلیوز ہوں؟ آپ `dropna` میں `how` اور `thresh` پیرامیٹرز کے ذریعے یہ سیٹنگز متعین کرتے ہیں۔

ڈیفالٹ طور پر، `how='any'` ہوتا ہے (اگر آپ خود چیک کرنا چاہتے ہیں یا دیکھنا چاہتے ہیں کہ اس میتھڈ کے دیگر پیرامیٹرز کیا ہیں، تو کوڈ سیل میں `example4.dropna?` چلائیں)۔ آپ متبادل طور پر `how='all'` متعین کر سکتے ہیں تاکہ صرف وہ قطاریں یا کالم ختم کیے جائیں جن میں تمام null ویلیوز ہوں۔ آئیے اپنے مثال کے `DataFrame` کو اگلی مشق میں دیکھتے ہیں تاکہ یہ عمل واضح ہو۔


In [25]:
example4[3] = np.nan
example4

0    1  2   3
0  1.0  NaN  7 NaN
1  2.0  5.0  8 NaN
2  NaN  6.0  9 NaN

> اہم نکات:
1. null ویلیوز کو ختم کرنا صرف تب ہی اچھا خیال ہے جب ڈیٹا سیٹ کافی بڑا ہو۔
2. مکمل قطاریں یا کالم ختم کیے جا سکتے ہیں اگر ان میں زیادہ تر ڈیٹا غائب ہو۔
3. `DataFrame.dropna(axis=)` میتھڈ null ویلیوز کو ختم کرنے میں مدد کرتا ہے۔ `axis` آرگومنٹ بتاتا ہے کہ قطاریں ختم کی جائیں یا کالم۔
4. `how` آرگومنٹ بھی استعمال کیا جا سکتا ہے۔ ڈیفالٹ میں یہ `any` پر سیٹ ہوتا ہے، اس لیے یہ صرف ان قطاروں/کالموں کو ختم کرتا ہے جن میں کوئی بھی null ویلیو ہو۔ اسے `all` پر سیٹ کیا جا سکتا ہے تاکہ صرف ان قطاروں/کالموں کو ختم کیا جائے جہاں تمام ویلیوز null ہوں۔


### ورزش:


In [22]:
# How might you go about dropping just column 3?
# Hint: remember that you will need to supply both the axis parameter and the how parameter.


`thresh` پیرامیٹر آپ کو زیادہ باریک بینی سے کنٹرول فراہم کرتا ہے: آپ اس تعداد کو مقرر کرتے ہیں کہ ایک قطار یا ستون میں کتنی *غیر-خالی* قدریں ہونی چاہئیں تاکہ اسے برقرار رکھا جا سکے۔


In [27]:
example4.dropna(axis='rows', thresh=3)

0    1  2   3
1  2.0  5.0  8 NaN

یہاں، پہلی اور آخری قطار کو حذف کر دیا گیا ہے، کیونکہ ان میں صرف دو غیر خالی اقدار موجود ہیں۔


### خالی قدروں کو بھرنا

کبھی کبھار خالی قدروں کو ایسی قدروں سے بھرنا مناسب ہوتا ہے جو ممکنہ طور پر درست ہو سکتی ہیں۔ خالی قدروں کو بھرنے کے لیے چند تکنیکیں موجود ہیں۔ پہلی تکنیک یہ ہے کہ موضوع کے علم (جس پر ڈیٹا سیٹ مبنی ہے) کا استعمال کرتے ہوئے خالی قدروں کا اندازہ لگایا جائے۔

آپ `isnull` کا استعمال کر کے یہ کام کر سکتے ہیں، لیکن یہ خاصا محنت طلب ہو سکتا ہے، خاص طور پر اگر آپ کے پاس بھرنے کے لیے بہت سی قدریں ہوں۔ چونکہ یہ کام ڈیٹا سائنس میں بہت عام ہے، pandas `fillna` فراہم کرتا ہے، جو `Series` یا `DataFrame` کی ایک کاپی واپس کرتا ہے جس میں خالی قدریں آپ کی منتخب کردہ قدر سے بدل دی جاتی ہیں۔ آئیے ایک اور مثال `Series` بناتے ہیں تاکہ دیکھ سکیں کہ یہ عملی طور پر کیسے کام کرتا ہے۔


### زمرہ بندی ڈیٹا (غیر عددی)
سب سے پہلے، آئیے غیر عددی ڈیٹا پر غور کریں۔ ڈیٹا سیٹس میں، ہمارے پاس زمرہ بندی ڈیٹا والے کالمز ہوتے ہیں۔ مثلاً جنس، صحیح یا غلط وغیرہ۔

زیادہ تر صورتوں میں، ہم گمشدہ اقدار کو کالم کے `mode` سے بدل دیتے ہیں۔ فرض کریں، ہمارے پاس 100 ڈیٹا پوائنٹس ہیں، جن میں سے 90 نے صحیح کہا، 8 نے غلط کہا اور 2 نے کچھ نہیں بھرا۔ تو ہم ان 2 کو صحیح سے بھر سکتے ہیں، پورے کالم کو مدنظر رکھتے ہوئے۔

یہاں بھی ہم ڈومین نالج استعمال کر سکتے ہیں۔ آئیے `mode` سے بھرنے کی ایک مثال پر غور کریں۔


In [28]:
fill_with_mode = pd.DataFrame([[1,2,"True"],
                               [3,4,None],
                               [5,6,"False"],
                               [7,8,"True"],
                               [9,10,"True"]])

fill_with_mode

0   1      2
0  1   2   True
1  3   4   None
2  5   6  False
3  7   8   True
4  9  10   True

اب، آئیے پہلے موڈ تلاش کریں اس سے پہلے کہ `None` کی قدر کو موڈ سے بھرا جائے۔


In [29]:
fill_with_mode[2].value_counts()

True     3
False    1
Name: 2, dtype: int64

تو، ہم None کو True سے بدل دیں گے۔


In [30]:
fill_with_mode[2].fillna('True',inplace=True)

In [31]:
fill_with_mode

0   1      2
0  1   2   True
1  3   4   True
2  5   6  False
3  7   8   True
4  9  10   True

جیسا کہ ہم دیکھ سکتے ہیں، خالی قدر کو تبدیل کر دیا گیا ہے۔ کہنے کی ضرورت نہیں، ہم اس جگہ کچھ بھی لکھ سکتے تھے یا `'True'` اور یہ تبدیل ہو جاتا۔


### عددی ڈیٹا
اب، عددی ڈیٹا کی بات کرتے ہیں۔ یہاں، گمشدہ اقدار کو تبدیل کرنے کے دو عام طریقے ہیں:

1. قطار کے میڈین سے تبدیل کریں
2. قطار کے اوسط سے تبدیل کریں

ہم میڈین سے تبدیل کرتے ہیں، جب ڈیٹا میں آؤٹ لائرز کے ساتھ اسکیو ہوتا ہے۔ اس کی وجہ یہ ہے کہ میڈین آؤٹ لائرز کے لیے مضبوط ہے۔

جب ڈیٹا نارملائز ہو، تو ہم اوسط استعمال کر سکتے ہیں، کیونکہ اس صورت میں اوسط اور میڈین کافی قریب ہوں گے۔

سب سے پہلے، آئیے ایک کالم لیتے ہیں جو عام طور پر تقسیم شدہ ہے اور گمشدہ قدر کو کالم کے اوسط سے بھرتے ہیں۔


In [32]:
fill_with_mean = pd.DataFrame([[-2,0,1],
                               [-1,2,3],
                               [np.nan,4,5],
                               [1,6,7],
                               [2,8,9]])

fill_with_mean

0  1  2
0 -2.0  0  1
1 -1.0  2  3
2  NaN  4  5
3  1.0  6  7
4  2.0  8  9

کالم کا اوسط ہے


In [33]:
np.mean(fill_with_mean[0])

0.0

میان سے بھرنا


In [34]:
fill_with_mean[0].fillna(np.mean(fill_with_mean[0]),inplace=True)
fill_with_mean

0  1  2
0 -2.0  0  1
1 -1.0  2  3
2  0.0  4  5
3  1.0  6  7
4  2.0  8  9

جیسا کہ ہم دیکھ سکتے ہیں، گمشدہ قدر کو اس کے اوسط سے بدل دیا گیا ہے۔


آئیے اب ایک اور ڈیٹا فریم آزمائیں، اور اس بار ہم None کی قدروں کو کالم کے درمیانی قدر سے بدل دیں گے۔


In [35]:
fill_with_median = pd.DataFrame([[-2,0,1],
                               [-1,2,3],
                               [0,np.nan,5],
                               [1,6,7],
                               [2,8,9]])

fill_with_median

0    1  2
0 -2  0.0  1
1 -1  2.0  3
2  0  NaN  5
3  1  6.0  7
4  2  8.0  9

دوسرے ستون کا میڈین ہے


In [36]:
fill_with_median[1].median()

4.0

میڈین کے ساتھ بھرنا


In [37]:
fill_with_median[1].fillna(fill_with_median[1].median(),inplace=True)
fill_with_median

0    1  2
0 -2  0.0  1
1 -1  2.0  3
2  0  4.0  5
3  1  6.0  7
4  2  8.0  9

جیسا کہ ہم دیکھ سکتے ہیں، NaN کی قدر کو کالم کے میڈین سے تبدیل کر دیا گیا ہے۔


In [38]:
example5 = pd.Series([1, np.nan, 2, None, 3], index=list('abcde'))
example5

a    1.0
b    NaN
c    2.0
d    NaN
e    3.0
dtype: float64

آپ تمام خالی اندراجات کو ایک واحد قدر، جیسے `0` سے بھر سکتے ہیں:


In [39]:
example5.fillna(0)

a    1.0
b    0.0
c    2.0
d    0.0
e    3.0
dtype: float64

> اہم نکات:
1. گمشدہ اقدار کو بھرنا تب ضروری ہوتا ہے جب یا تو ڈیٹا کم ہو یا گمشدہ ڈیٹا کو بھرنے کے لیے کوئی حکمت عملی موجود ہو۔
2. گمشدہ اقدار کو اندازے سے بھرنے کے لیے ڈومین علم کا استعمال کیا جا سکتا ہے۔
3. زمرہ جاتی ڈیٹا کے لیے، زیادہ تر گمشدہ اقدار کو کالم کے موڈ سے تبدیل کیا جاتا ہے۔
4. عددی ڈیٹا کے لیے، گمشدہ اقدار کو عام طور پر کالمز کے اوسط (نارملائزڈ ڈیٹا سیٹس کے لیے) یا میڈین سے بھرا جاتا ہے۔


### ورزش:


In [40]:
# What happens if you try to fill null values with a string, like ''?


آپ null قدروں کو **آگے بھر** سکتے ہیں، یعنی آخری درست قدر کو استعمال کرتے ہوئے null کو بھرنا:


In [41]:
example5.fillna(method='ffill')

a    1.0
b    1.0
c    2.0
d    2.0
e    3.0
dtype: float64

آپ اگلے درست قدر کو پیچھے کی طرف پھیلانے کے لیے **بیک فل** بھی کر سکتے ہیں تاکہ ایک نل کو بھر سکیں۔


In [42]:
example5.fillna(method='bfill')

a    1.0
b    2.0
c    2.0
d    3.0
e    3.0
dtype: float64

جیسا کہ آپ اندازہ لگا سکتے ہیں، یہ ڈیٹا فریمز کے ساتھ بھی اسی طرح کام کرتا ہے، لیکن آپ ایک `axis` بھی مخصوص کر سکتے ہیں جس کے ساتھ null قدروں کو بھرنا ہے۔


In [43]:
example4

0    1  2   3
0  1.0  NaN  7 NaN
1  2.0  5.0  8 NaN
2  NaN  6.0  9 NaN

In [44]:
example4.fillna(method='ffill', axis=1)

0    1    2    3
0  1.0  1.0  7.0  7.0
1  2.0  5.0  8.0  8.0
2  NaN  6.0  9.0  9.0

نوٹ کریں کہ جب فارورڈ فلنگ کے لیے پچھلی قدر دستیاب نہیں ہوتی، تو خالی قدر برقرار رہتی ہے۔


### ورزش:


In [45]:
# What output does example4.fillna(method='bfill', axis=1) produce?
# What about example4.fillna(method='ffill') or example4.fillna(method='bfill')?
# Can you think of a longer code snippet to write that can fill all of the null values in example4?


آپ `fillna` کے استعمال میں تخلیقی ہو سکتے ہیں۔ مثال کے طور پر، آئیے دوبارہ `example4` کو دیکھتے ہیں، لیکن اس بار آئیے گمشدہ قدروں کو `DataFrame` میں موجود تمام قدروں کے اوسط سے بھر دیتے ہیں:


In [46]:
example4.fillna(example4.mean())

0    1  2   3
0  1.0  5.5  7 NaN
1  2.0  5.0  8 NaN
2  1.5  6.0  9 NaN

نوٹ کریں کہ کالم 3 ابھی بھی خالی ہے: ڈیفالٹ سمت یہ ہے کہ اقدار کو قطار کے لحاظ سے بھرا جائے۔

> **اہم نکتہ:** اپنے ڈیٹا سیٹس میں موجود خالی اقدار سے نمٹنے کے کئی طریقے ہیں۔ جو خاص حکمت عملی آپ اپناتے ہیں (انہیں ہٹانا، تبدیل کرنا، یا یہاں تک کہ کس طرح تبدیل کرنا) وہ ڈیٹا کی خصوصیات پر منحصر ہونا چاہیے۔ جتنا زیادہ آپ ڈیٹا سیٹس کے ساتھ کام کریں گے اور ان سے تعامل کریں گے، اتنا ہی بہتر آپ خالی اقدار سے نمٹنے کا طریقہ سمجھیں گے۔


### زمرہ وار ڈیٹا کو انکوڈ کرنا

مشین لرننگ ماڈلز صرف نمبروں اور کسی بھی قسم کے عددی ڈیٹا کے ساتھ کام کرتے ہیں۔ یہ "ہاں" اور "نہیں" کے درمیان فرق نہیں بتا سکتے، لیکن یہ 0 اور 1 کے درمیان فرق کر سکتے ہیں۔ لہذا، گمشدہ اقدار کو بھرنے کے بعد، ہمیں زمرہ وار ڈیٹا کو کسی عددی شکل میں انکوڈ کرنا ہوگا تاکہ ماڈل اسے سمجھ سکے۔

انکوڈنگ دو طریقوں سے کی جا سکتی ہے۔ ہم اگلے حصے میں ان پر بات کریں گے۔


**لیبل انکوڈنگ**

لیبل انکوڈنگ بنیادی طور پر ہر زمرے کو ایک نمبر میں تبدیل کرنا ہے۔ مثال کے طور پر، فرض کریں کہ ہمارے پاس ایئر لائن مسافروں کا ایک ڈیٹا سیٹ ہے اور اس میں ایک کالم ہے جو ان کی کلاس کو ظاہر کرتا ہے، جیسے ['بزنس کلاس', 'اکانومی کلاس', 'فرسٹ کلاس']۔ اگر اس پر لیبل انکوڈنگ کی جائے تو یہ [0,1,2] میں تبدیل ہو جائے گا۔ آئیے کوڈ کے ذریعے ایک مثال دیکھتے ہیں۔ چونکہ ہم آنے والے نوٹ بکس میں `scikit-learn` سیکھیں گے، اس لیے ہم اسے یہاں استعمال نہیں کریں گے۔


In [47]:
label = pd.DataFrame([
                      [10,'business class'],
                      [20,'first class'],
                      [30, 'economy class'],
                      [40, 'economy class'],
                      [50, 'economy class'],
                      [60, 'business class']
],columns=['ID','class'])
label

ID           class
0  10  business class
1  20     first class
2  30   economy class
3  40   economy class
4  50   economy class
5  60  business class

پہلے کالم پر لیبل انکوڈنگ کرنے کے لیے، ہمیں پہلے ہر کلاس سے ایک نمبر تک نقشہ بندی بیان کرنی ہوگی، اس سے پہلے کہ ہم تبدیل کریں۔


In [48]:
class_labels = {'business class':0,'economy class':1,'first class':2}
label['class'] = label['class'].replace(class_labels)
label

ID  class
0  10      0
1  20      2
2  30      1
3  40      1
4  50      1
5  60      0

جیسا کہ ہم دیکھ سکتے ہیں، نتیجہ وہی ہے جو ہم نے سوچا تھا کہ ہوگا۔ تو، لیبل انکوڈنگ کب استعمال کی جاتی ہے؟ لیبل انکوڈنگ درج ذیل میں سے کسی ایک یا دونوں صورتوں میں استعمال کی جاتی ہے:
1. جب زمروں کی تعداد زیادہ ہو
2. جب زمرے ترتیب میں ہوں۔


**ون ہاٹ انکوڈنگ**

ون ہاٹ انکوڈنگ ایک اور قسم کی انکوڈنگ ہے۔ اس قسم کی انکوڈنگ میں، کالم کی ہر کیٹیگری کو ایک الگ کالم کے طور پر شامل کیا جاتا ہے اور ہر ڈیٹا پوائنٹ کو 0 یا 1 دیا جاتا ہے، اس بنیاد پر کہ آیا وہ کیٹیگری موجود ہے یا نہیں۔ لہذا، اگر n مختلف کیٹیگریز ہوں، تو n کالمز ڈیٹا فریم میں شامل کیے جائیں گے۔

مثال کے طور پر، آئیے وہی ہوائی جہاز کی کلاس کی مثال لیتے ہیں۔ کیٹیگریز تھیں: ['بزنس کلاس', 'اکانومی کلاس', 'فرسٹ کلاس']۔ تو، اگر ہم ون ہاٹ انکوڈنگ کریں، تو درج ذیل تین کالمز ڈیٹا سیٹ میں شامل کیے جائیں گے: ['class_business class', 'class_economy class', 'class_first class']۔


In [49]:
one_hot = pd.DataFrame([
                      [10,'business class'],
                      [20,'first class'],
                      [30, 'economy class'],
                      [40, 'economy class'],
                      [50, 'economy class'],
                      [60, 'business class']
],columns=['ID','class'])
one_hot

ID           class
0  10  business class
1  20     first class
2  30   economy class
3  40   economy class
4  50   economy class
5  60  business class

آئیے پہلی کالم پر ون ہاٹ انکوڈنگ انجام دیں۔


In [50]:
one_hot_data = pd.get_dummies(one_hot,columns=['class'])

In [51]:
one_hot_data

ID  class_business class  class_economy class  class_first class
0  10                     1                    0                  0
1  20                     0                    0                  1
2  30                     0                    1                  0
3  40                     0                    1                  0
4  50                     0                    1                  0
5  60                     1                    0                  0

ہر ایک ہاٹ انکوڈڈ کالم میں 0 یا 1 ہوتا ہے، جو یہ ظاہر کرتا ہے کہ آیا وہ زمرہ اس ڈیٹا پوائنٹ کے لیے موجود ہے۔


ہم ایک ہاٹ انکوڈنگ کب استعمال کرتے ہیں؟ ایک ہاٹ انکوڈنگ درج ذیل میں سے ایک یا دونوں صورتوں میں استعمال کی جاتی ہے:

1. جب زمروں کی تعداد اور ڈیٹا سیٹ کا سائز چھوٹا ہو۔
2. جب زمرے کسی خاص ترتیب کی پیروی نہ کرتے ہوں۔


> اہم نکات:
1. انکوڈنگ غیر عددی ڈیٹا کو عددی ڈیٹا میں تبدیل کرنے کے لیے کی جاتی ہے۔
2. انکوڈنگ کی دو اقسام ہیں: لیبل انکوڈنگ اور ون ہاٹ انکوڈنگ، جنہیں ڈیٹا سیٹ کی ضروریات کے مطابق انجام دیا جا سکتا ہے۔


## ڈپلیکیٹ ڈیٹا کو ہٹانا

> **سیکھنے کا مقصد:** اس ذیلی حصے کے اختتام تک، آپ کو ڈیٹا فریمز سے ڈپلیکیٹ ویلیوز کی شناخت اور انہیں ہٹانے میں مہارت حاصل ہو جانی چاہیے۔

گمشدہ ڈیٹا کے علاوہ، آپ کو حقیقی دنیا کے ڈیٹا سیٹس میں اکثر ڈپلیکیٹ ڈیٹا کا سامنا کرنا پڑے گا۔ خوش قسمتی سے، pandas ڈپلیکیٹ اندراجات کا پتہ لگانے اور انہیں ہٹانے کا ایک آسان طریقہ فراہم کرتا ہے۔


### ڈپلیکیٹس کی شناخت: `duplicated`

آپ pandas میں `duplicated` میتھڈ کا استعمال کرکے آسانی سے ڈپلیکیٹ ویلیوز کو پہچان سکتے ہیں، جو ایک Boolean ماسک واپس کرتا ہے جو یہ ظاہر کرتا ہے کہ آیا `DataFrame` میں کوئی اندراج پہلے والے کا ڈپلیکیٹ ہے۔ آئیے ایک اور مثال کے طور پر `DataFrame` بناتے ہیں تاکہ اسے عملی طور پر دیکھ سکیں۔


In [52]:
example6 = pd.DataFrame({'letters': ['A','B'] * 2 + ['B'],
                         'numbers': [1, 2, 1, 3, 3]})
example6

letters  numbers
0       A        1
1       B        2
2       A        1
3       B        3
4       B        3

In [53]:
example6.duplicated()

0    False
1    False
2     True
3    False
4     True
dtype: bool

### ڈپلیکیٹس کو ختم کرنا: `drop_duplicates`
`drop_duplicates` صرف ان ڈیٹا کی ایک کاپی واپس کرتا ہے جن کے تمام `duplicated` ویلیوز `False` ہیں:


In [54]:
example6.drop_duplicates()

letters  numbers
0       A        1
1       B        2
3       B        3

`duplicated` اور `drop_duplicates` دونوں ڈیفالٹ طور پر تمام کالمز کو مدنظر رکھتے ہیں لیکن آپ یہ مخصوص کر سکتے ہیں کہ وہ آپ کے `DataFrame` میں صرف کالمز کے ایک ذیلی سیٹ کا جائزہ لیں:


In [55]:
example6.drop_duplicates(['letters'])

letters  numbers
0       A        1
1       B        2

> **حاصلِ کلام:** نقل شدہ ڈیٹا کو ہٹانا تقریباً ہر ڈیٹا سائنس منصوبے کا ایک لازمی حصہ ہے۔ نقل شدہ ڈیٹا آپ کے تجزیوں کے نتائج کو بدل سکتا ہے اور آپ کو غلط نتائج دے سکتا ہے!


## حقیقی دنیا کے ڈیٹا کے معیار کی جانچ

> **سیکھنے کا مقصد:** اس حصے کے اختتام تک، آپ حقیقی دنیا کے ڈیٹا کے معیار کے عام مسائل جیسے غیر مستقل زمرہ جاتی اقدار، غیر معمولی عددی اقدار (آؤٹ لائرز)، اور معمولی فرق کے ساتھ نقل شدہ اداروں کا پتہ لگانے اور درست کرنے میں مہارت حاصل کر لیں گے۔

اگرچہ گمشدہ اقدار اور بالکل نقل شدہ ڈیٹا عام مسائل ہیں، حقیقی دنیا کے ڈیٹا سیٹس میں اکثر زیادہ پیچیدہ مسائل شامل ہوتے ہیں:

1. **غیر مستقل زمرہ جاتی اقدار**: ایک ہی زمرہ مختلف طریقوں سے لکھا گیا (جیسے "USA"، "U.S.A"، "United States")
2. **غیر معمولی عددی اقدار**: انتہائی آؤٹ لائرز جو ڈیٹا انٹری کی غلطیوں کی نشاندہی کرتے ہیں (جیسے عمر = 999)
3. **قریب قریب نقل شدہ قطاریں**: وہ ریکارڈز جو ایک ہی ادارے کی نمائندگی کرتے ہیں لیکن معمولی فرق کے ساتھ

آئیے ان مسائل کا پتہ لگانے اور ان سے نمٹنے کے طریقے دریافت کرتے ہیں۔


### ایک نمونہ "گندا" ڈیٹاسیٹ بنانا

سب سے پہلے، آئیے ایک نمونہ ڈیٹاسیٹ بناتے ہیں جس میں وہ مسائل شامل ہوں جو ہم عام طور پر حقیقی دنیا کے ڈیٹا میں دیکھتے ہیں:


In [ ]:
import pandas as pd
import numpy as np

# Create a sample dataset with quality issues
dirty_data = pd.DataFrame({
    'customer_id': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12],
    'name': ['John Smith', 'Jane Doe', 'John Smith', 'Bob Johnson', 
             'Alice Williams', 'Charlie Brown', 'John  Smith', 'Eva Martinez',
             'Bob Johnson', 'Diana Prince', 'Frank Castle', 'Alice Williams'],
    'age': [25, 32, 25, 45, 28, 199, 25, 31, 45, 27, -5, 28],
    'country': ['USA', 'UK', 'U.S.A', 'Canada', 'USA', 'United Kingdom',
                'United States', 'Mexico', 'canada', 'USA', 'UK', 'usa'],
    'purchase_amount': [100.50, 250.00, 105.00, 320.00, 180.00, 90.00,
                       102.00, 275.00, 325.00, 195.00, 410.00, 185.00]
})

print("Sample 'Dirty' Dataset:")
print(dirty_data)

### 1. غیر متوازن زمرہ جاتی اقدار کا پتہ لگانا

دھیان دیں کہ `country` کالم میں ایک ہی ممالک کے لیے متعدد نمائندگیاں موجود ہیں۔ آئیے ان بے ترتیبیوں کی نشاندہی کرتے ہیں:


In [ ]:
# Check unique values in the country column
print("Unique country values:")
print(dirty_data['country'].unique())
print(f"\nTotal unique values: {dirty_data['country'].nunique()}")

# Count occurrences of each variation
print("\nValue counts:")
print(dirty_data['country'].value_counts())

#### زمرہ وار اقدار کو معیاری بنانا

ہم ان اقدار کو معیاری بنانے کے لیے ایک نقشہ بنا سکتے ہیں۔ ایک آسان طریقہ یہ ہے کہ انہیں چھوٹے حروف میں تبدیل کریں اور ایک نقشہ سازی کی ڈکشنری بنائیں:


In [ ]:
# Create a standardization mapping
country_mapping = {
    'usa': 'USA',
    'u.s.a': 'USA',
    'united states': 'USA',
    'uk': 'UK',
    'united kingdom': 'UK',
    'canada': 'Canada',
    'mexico': 'Mexico'
}

# Standardize the country column
dirty_data['country_clean'] = dirty_data['country'].str.lower().map(country_mapping)

print("Before standardization:")
print(dirty_data['country'].value_counts())
print("\nAfter standardization:")
print(dirty_data[['country_clean']].value_counts())

**متبادل: دھندلے مماثلت کا استعمال**

زیادہ پیچیدہ معاملات کے لیے، ہم `rapidfuzz` لائبریری کے ساتھ دھندلے سٹرنگ مماثلت کا استعمال کر سکتے ہیں تاکہ خودکار طور پر مشابہت رکھنے والے سٹرنگز کا پتہ لگایا جا سکے:


In [ ]:
try:
    from rapidfuzz import process, fuzz
except ImportError:
    print("rapidfuzz is not installed. Please install it with 'pip install rapidfuzz' to use fuzzy matching.")
    process = None
    fuzz = None

# Get unique countries
unique_countries = dirty_data['country'].unique()

# For each country, find similar matches
if process is not None and fuzz is not None:
    print("Finding similar country names (similarity > 70%):")
    for country in unique_countries:
        matches = process.extract(country, unique_countries, scorer=fuzz.ratio, limit=3)
        # Filter matches with similarity > 70 and not identical
        similar = [m for m in matches if m[1] > 70 and m[0] != country]
        if similar:
            print(f"\n'{country}' is similar to:")
            for match, score, _ in similar:
                print(f"  - '{match}' (similarity: {score}%)")
else:
    print("Skipping fuzzy matching because rapidfuzz is not available.")

### 2. غیر معمولی عددی قدروں کا پتہ لگانا (آؤٹ لائرز)

`age` کالم کو دیکھتے ہوئے، ہمارے پاس کچھ مشکوک قدریں ہیں جیسے 199 اور -5۔ آئیے ان آؤٹ لائرز کا پتہ لگانے کے لیے شماریاتی طریقے استعمال کریں۔


In [ ]:
# Display basic statistics
print("Age column statistics:")
print(dirty_data['age'].describe())

# Identify impossible values using domain knowledge
print("\nRows with impossible age values (< 0 or > 120):")
impossible_ages = dirty_data[(dirty_data['age'] < 0) | (dirty_data['age'] > 120)]
print(impossible_ages[['customer_id', 'name', 'age']])

#### آئی کیو آر (Interquartile Range) طریقہ استعمال کرنا

آئی کیو آر طریقہ ایک مضبوط شماریاتی تکنیک ہے جو غیر معمولی اقدار کی شناخت کے لیے استعمال ہوتی ہے اور انتہائی اقدار کے لیے کم حساس ہے:


In [ ]:
# Calculate IQR for age (excluding impossible values)
valid_ages = dirty_data[(dirty_data['age'] >= 0) & (dirty_data['age'] <= 120)]['age']

Q1 = valid_ages.quantile(0.25)
Q3 = valid_ages.quantile(0.75)
IQR = Q3 - Q1

# Define outlier bounds
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

print(f"IQR-based outlier bounds for age: [{lower_bound:.2f}, {upper_bound:.2f}]")

# Identify outliers
age_outliers = dirty_data[(dirty_data['age'] < lower_bound) | (dirty_data['age'] > upper_bound)]
print(f"\nRows with age outliers:")
print(age_outliers[['customer_id', 'name', 'age']])

#### زیڈ اسکور طریقہ استعمال کرنا

زیڈ اسکور طریقہ اوسط سے معیاری انحرافات کی بنیاد پر غیر معمولی اقدار کی نشاندہی کرتا ہے:


In [ ]:
try:
    from scipy import stats
except ImportError:
    print("scipy is required for Z-score calculation. Please install it with 'pip install scipy' and rerun this cell.")
else:
    # Calculate Z-scores for age, handling NaN values
    age_nonan = dirty_data['age'].dropna()
    zscores = np.abs(stats.zscore(age_nonan))
    dirty_data['age_zscore'] = np.nan
    dirty_data.loc[age_nonan.index, 'age_zscore'] = zscores

    # Typically, Z-score > 3 indicates an outlier
    print("Rows with age Z-score > 3:")
    zscore_outliers = dirty_data[dirty_data['age_zscore'] > 3]
    print(zscore_outliers[['customer_id', 'name', 'age', 'age_zscore']])

    # Clean up the temporary column
    dirty_data = dirty_data.drop('age_zscore', axis=1)

#### آؤٹ لائرز کو سنبھالنا

جب آؤٹ لائرز کا پتہ چل جائے، تو انہیں کئی طریقوں سے سنبھالا جا سکتا ہے:
1. **ہٹائیں**: آؤٹ لائرز والی قطاریں حذف کریں (اگر وہ غلطیاں ہیں)
2. **حد مقرر کریں**: حد کے اقدار سے تبدیل کریں
3. **NaN سے تبدیل کریں**: انہیں گمشدہ ڈیٹا کے طور پر سمجھیں اور امپیوٹیشن تکنیک استعمال کریں
4. **رکھیں**: اگر وہ جائز انتہائی اقدار ہیں


In [ ]:
# Create a cleaned version by replacing impossible ages with NaN
dirty_data['age_clean'] = dirty_data['age'].apply(
    lambda x: np.nan if (x < 0 or x > 120) else x
)

print("Age column before and after cleaning:")
print(dirty_data[['customer_id', 'name', 'age', 'age_clean']])

### 3. قریب المثل قطاروں کا پتہ لگانا

دھیان دیں کہ ہمارے ڈیٹا سیٹ میں "John Smith" کے لیے کئی اندراجات موجود ہیں جن میں قدرے مختلف اقدار ہیں۔ آئیے نام کی مماثلت کی بنیاد پر ممکنہ نقلیں شناخت کریں۔


In [ ]:
# First, let's look at exact name matches (ignoring extra whitespace)
dirty_data['name_normalized'] = dirty_data['name'].str.strip().str.lower()

print("Checking for duplicate names:")
duplicate_names = dirty_data[dirty_data.duplicated(['name_normalized'], keep=False)]
print(duplicate_names.sort_values('name_normalized')[['customer_id', 'name', 'age', 'country']])

#### دھندلے مماثلت کے ساتھ قریباً نقلیں تلاش کرنا

زیادہ پیچیدہ نقلیں تلاش کرنے کے لیے، ہم دھندلے مماثلت کا استعمال کر سکتے ہیں تاکہ ملتے جلتے نام تلاش کیے جا سکیں:


In [ ]:
try:
    from rapidfuzz import process, fuzz

    # Function to find potential duplicates
    def find_near_duplicates(df, column, threshold=90):
        """
        Find near-duplicate entries in a column using fuzzy matching.
        
        Parameters:
        - df: DataFrame
        - column: Column name to check for duplicates
        - threshold: Similarity threshold (0-100)
        
        Returns: List of potential duplicate groups
        """
        values = df[column].unique()
        duplicate_groups = []
        checked = set()
        
        for value in values:
            if value in checked:
                continue
                
            # Find similar values
            matches = process.extract(value, values, scorer=fuzz.ratio, limit=len(values))
            similar = [m[0] for m in matches if m[1] >= threshold]
            
            if len(similar) > 1:
                duplicate_groups.append(similar)
                checked.update(similar)
        
        return duplicate_groups

    # Find near-duplicate names
    duplicate_groups = find_near_duplicates(dirty_data, 'name', threshold=90)

    print("Potential duplicate groups:")
    for i, group in enumerate(duplicate_groups, 1):
        print(f"\nGroup {i}:")
        for name in group:
            matching_rows = dirty_data[dirty_data['name'] == name]
            print(f"  '{name}': {len(matching_rows)} occurrence(s)")
            for _, row in matching_rows.iterrows():
                print(f"    - Customer {row['customer_id']}: age={row['age']}, country={row['country']}")
except ImportError:
    print("rapidfuzz is not installed. Skipping fuzzy matching for near-duplicates.")

#### ڈپلیکیٹس کو ہینڈل کرنا

جب ڈپلیکیٹس کی شناخت ہو جائے، تو آپ کو فیصلہ کرنا ہوگا کہ انہیں کیسے ہینڈل کریں:
1. **پہلی بار وقوع کو رکھیں**: `drop_duplicates(keep='first')` استعمال کریں
2. **آخری بار وقوع کو رکھیں**: `drop_duplicates(keep='last')` استعمال کریں
3. **معلومات کو جمع کریں**: ڈپلیکیٹ قطاروں سے معلومات کو یکجا کریں
4. **دستی جائزہ**: انسانی جائزے کے لیے نشان زد کریں


In [ ]:
# Example: Remove duplicates based on normalized name, keeping first occurrence
cleaned_data = dirty_data.drop_duplicates(subset=['name_normalized'], keep='first')

print(f"Original dataset: {len(dirty_data)} rows")
print(f"After removing name duplicates: {len(cleaned_data)} rows")
print(f"Removed: {len(dirty_data) - len(cleaned_data)} duplicate rows")

print("\nCleaned dataset:")
print(cleaned_data[['customer_id', 'name', 'age', 'country_clean']])

### خلاصہ: مکمل ڈیٹا صفائی کا عمل

آئیے اسے ایک جامع صفائی کے عمل میں یکجا کرتے ہیں:


In [ ]:
def clean_dataset(df):
    """
    Comprehensive data cleaning function.
    """
    # Create a copy to avoid modifying the original
    cleaned = df.copy()
    
    # 1. Standardize categorical values (country)
    country_mapping = {
        'usa': 'USA', 'u.s.a': 'USA', 'united states': 'USA',
        'uk': 'UK', 'united kingdom': 'UK',
        'canada': 'Canada', 'mexico': 'Mexico'
    }
    cleaned['country'] = cleaned['country'].str.lower().map(country_mapping)
    
    # 2. Clean abnormal age values
    cleaned['age'] = cleaned['age'].apply(
        lambda x: np.nan if (x < 0 or x > 120) else x
    )
    
    # 3. Remove near-duplicate names (normalize whitespace)
    cleaned['name'] = cleaned['name'].str.strip()
    cleaned = cleaned.drop_duplicates(subset=['name'], keep='first')
    
    return cleaned

# Apply the cleaning pipeline
final_cleaned_data = clean_dataset(dirty_data)

print("Before cleaning:")
print(f"  Rows: {len(dirty_data)}")
print(f"  Unique countries: {dirty_data['country'].nunique()}")
print(f"  Invalid ages: {((dirty_data['age'] < 0) | (dirty_data['age'] > 120)).sum()}")

print("\nAfter cleaning:")
print(f"  Rows: {len(final_cleaned_data)}")
print(f"  Unique countries: {final_cleaned_data['country'].nunique()}")
print(f"  Invalid ages: {((final_cleaned_data['age'] < 0) | (final_cleaned_data['age'] > 120)).sum()}")

print("\nCleaned dataset:")
print(final_cleaned_data[['customer_id', 'name', 'age', 'country', 'purchase_amount']])

### 🎯 چیلنج مشق

اب یہ آپ کی باری ہے! نیچے ایک نئی قطار دی گئی ہے جس میں کئی معیار کے مسائل ہیں۔ کیا آپ:

1. اس قطار میں موجود تمام مسائل کی نشاندہی کر سکتے ہیں؟
2. ہر مسئلے کو صاف کرنے کے لیے کوڈ لکھ سکتے ہیں؟
3. صاف شدہ قطار کو ڈیٹاسیٹ میں شامل کر سکتے ہیں؟

یہاں مسئلہ زدہ ڈیٹا ہے:


In [ ]:
# New problematic row
new_row = pd.DataFrame({
    'customer_id': [13],
    'name': ['  Diana  Prince  '],  # Extra whitespace
    'age': [250],  # Impossible age
    'country': ['U.S.A.'],  # Inconsistent format
    'purchase_amount': [150.00]
})

print("New row to clean:")
print(new_row)

# TODO: Your code here to clean this row
# Hints:
# 1. Strip whitespace from the name
# 2. Check if the name is a duplicate (Diana Prince already exists)
# 3. Handle the impossible age value
# 4. Standardize the country name

# Example solution (uncomment and modify as needed):
# new_row_cleaned = new_row.copy()
# new_row_cleaned['name'] = new_row_cleaned['name'].str.strip()
# new_row_cleaned['age'] = np.nan  # Invalid age
# new_row_cleaned['country'] = 'USA'  # Standardized
# print("\nCleaned row:")
# print(new_row_cleaned)

### اہم نکات

1. **غیر مستقل زمروں** کا حقیقی دنیا کے ڈیٹا میں عام ہونا معمول کی بات ہے۔ ہمیشہ منفرد اقدار کو چیک کریں اور انہیں میپنگ یا فزی میچنگ کے ذریعے معیاری بنائیں۔

2. **آؤٹ لائرز** آپ کے تجزیے پر نمایاں اثر ڈال سکتے ہیں۔ انہیں شناخت کرنے کے لیے ڈومین نالج کو شماریاتی طریقوں (IQR، Z-score) کے ساتھ استعمال کریں۔

3. **قریب قریب نقلیں** بالکل نقلوں کی نسبت زیادہ مشکل ہوتی ہیں۔ انہیں پہچاننے کے لیے فزی میچنگ اور ڈیٹا کو نارملائز کرنے (لوئر کیس کرنا، اسپیس ہٹانا) پر غور کریں۔

4. **ڈیٹا کی صفائی ایک تکراری عمل ہے**۔ آپ کو کئی تکنیکیں استعمال کرنی پڑ سکتی ہیں اور نتائج کا جائزہ لینا پڑ سکتا ہے، اس سے پہلے کہ آپ اپنے صاف شدہ ڈیٹا سیٹ کو حتمی شکل دیں۔

5. **اپنے فیصلوں کو دستاویزی بنائیں**۔ یہ یاد رکھیں کہ آپ نے کون سے صفائی کے اقدامات کیے اور کیوں، کیونکہ یہ دوبارہ قابل عمل اور شفافیت کے لیے اہم ہے۔

> **بہترین طریقہ:** ہمیشہ اپنے اصل "گندے" ڈیٹا کی ایک کاپی محفوظ رکھیں۔ اپنے سورس ڈیٹا فائلز کو کبھی اوور رائٹ نہ کریں - صاف شدہ ورژنز بنائیں اور واضح نام دینے کے اصول اپنائیں جیسے `data_cleaned.csv`۔



---

**ڈسکلیمر**:  
یہ دستاویز AI ترجمہ سروس [Co-op Translator](https://github.com/Azure/co-op-translator) کا استعمال کرتے ہوئے ترجمہ کی گئی ہے۔ ہم درستگی کے لیے کوشش کرتے ہیں، لیکن براہ کرم آگاہ رہیں کہ خودکار ترجمے میں غلطیاں یا غیر درستیاں ہو سکتی ہیں۔ اصل دستاویز کو اس کی اصل زبان میں مستند ذریعہ سمجھا جانا چاہیے۔ اہم معلومات کے لیے، پیشہ ور انسانی ترجمہ کی سفارش کی جاتی ہے۔ ہم اس ترجمے کے استعمال سے پیدا ہونے والی کسی بھی غلط فہمی یا غلط تشریح کے ذمہ دار نہیں ہیں۔
